### PR TITLE
Add `Precompiled header` for modules & Refine by PImpl to accelerate release build

### PR DESCRIPTION
**Pull request by solotzg**
### What problem does this PR solve?

Issue Number: ref #4909

Problem Summary:

Ref: https://solotzg.github.io/2022/06/01/system/cpp-compile-optimize/#%E9%A2%84%E7%BC%96%E8%AF%91%E5%A4%B4%E6%96%87%E4%BB%B6%EF%BC%88PCH%EF%BC%89

Before
```
**** Time summary:
Compilation (5485 times):
  Parsing (frontend):         7181.8 s
  Codegen & opts (backend):   6224.5 s

*** Expensive headers:
1091837 ms: /root/work/tiflash/dbms/src/Common/FmtUtils.h (included 750 times, avg 1455 ms), included via:
1022467 ms: /root/work/tiflash/libs/libcommon/include/common/StringRef.h (included 754 times, avg 1356 ms), included via:
985876 ms: /root/work/tiflash/dbms/src/Common/Exception.h (included 748 times, avg 1318 ms), included via:
974680 ms: /root/work/tiflash/dbms/src/Common/StackTrace.h (included 749 times, avg 1301 ms), included via:
681167 ms: /data2/tiflash-env/sysroot/include/c++/v1/string (included 1903 times, avg 357 ms), included via:
501559 ms: /data2/tiflash-env/sysroot/include/c++/v1/algorithm (included 1969 times, avg 254 ms), included via:
498058 ms: /root/work/tiflash/dbms/src/Columns/IColumn.h (included 516 times, avg 965 ms), included via:
422329 ms: /root/work/tiflash/dbms/src/Core/Types.h (included 706 times, avg 598 ms), included via:
405688 ms: /data2/tiflash-env/sysroot/include/c++/v1/functional (included 1971 times, avg 205 ms), included via:
```

After
```
**** Time summary:
Compilation (5488 times):
  Parsing (frontend):         4208.2 s
  Codegen & opts (backend):   6663.9 s

*** Expensive headers:
188325 ms: /data2/work/tiflash/dbms/src/Columns/IColumn.h (included 515 times, avg 365 ms), included via:
...
177991 ms: /data2/work/tiflash/dbms/src/Common/Decimal.h (included 657 times, avg 270 ms), included via:
...
122434 ms: /data2/work/tiflash/dbms/src/Core/Block.h (included 415 times, avg 295 ms), included via:
...
115633 ms: /data2/work/tiflash/dbms/src/Core/Field.h (included 656 times, avg 176 ms), included via:
...
```

With the help of LTO(#4890, #4924), [Forward declaration](https://en.wikipedia.org/wiki/Forward_declaration) won't hurt performance because linker(lld) will make short functions inline automatically.

For example: use `objdump -d`, there will be only one entry to function `PlusDecimalInferer::infer` in code section. There is no other function call towards `PlusDecimalInferer::infer` by instruction `callq`.

```
➜  release-centos7-llvm git:(pch) ✗ ./tiflash/tiflash version
TiFlash
Release Version: v6.2.0-alpha-30-gf2a406a
Edition:         Community
Git Commit Hash: f2a406aa93f06e98e1868597f0ff25ae87ed40c5
Git Branch:      pch
UTC Build Time:  2022-06-05 07:30:43
Enable Features: jemalloc avx avx512 unwind thinlto
Profile:         RELWITHDEBINFO

Raft Proxy
Git Commit Hash:   ca2f51f94e55bdd23749dcc02ab4afb94eeb5ae5
Git Commit Branch: HEAD
UTC Build Time:    2022-06-05 07:33:25
Rust Version:      rustc 1.60.0-nightly (1e12aef3f 2022-02-13)
Storage Engine:    tiflash
Prometheus Prefix: tiflash_proxy_
Profile:           release

➜  release-centos7-llvm git:(pch) ✗ objdump -d ./tiflash/tiflash | grep 'PlusDecimalInferer'
0000000001ce43c0 <_ZN2DB18PlusDecimalInferer5inferEjjjj>:```

➜  release-centos7-llvm git:(pch) ✗ objdump -d ./tiflash/tiflash | grep '1ce43c0'           
0000000001ce43c0 <_ZN2DB18PlusDecimalInferer5inferEjjjj>:
 1ce43c0:       55                      push   %rbp
```

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

**Review by SchrodingerZhu — APPROVED** [review]

File: dbms/src/Common/Decimal.h [hunk 1]
```diff
@@ -108,66 +108,40 @@ struct IntPrec<Int256>
...
 struct PlusDecimalInferer
 {
-    static std::tuple<PrecType, ScaleType> infer(PrecType left_prec, ScaleType left_scale, PrecType right_prec, ScaleType right_scale)
-    {
-        ScaleType result_scale = std::max(left_scale, right_scale);
-        PrecType result_int = std::max(left_prec - left_scale, right_prec - right_scale);
-        PrecType result_prec = std::min(result_scale + result_int + 1, decimal_max_prec);
-        return {result_prec, result_scale};
-    }
+    static std::tuple<PrecType, ScaleType> infer(PrecType left_prec, ScaleType left_scale, PrecType right_prec, ScaleType right_scale);
```
Comment (fuzhe1989) [inline dbms/src/Common/Decimal.h:111]: This pr moves lost of implementations of inline functions into cpp files, so they are never inlineable, which could hurt the performance, is that what you expect?

Comment (solotzg) [reply dbms/src/Common/Decimal.h:111]: In https://github.com/pingcap/tiflash/pull/4890 and https://github.com/pingcap/tiflash/pull/4924, we have enabled LTO  in release build process. Compiler will inline those short methods automatically.

For example: use `objdump -d`, there will be only entry to function in code section. There is no other function call towards `PlusDecimalInferer::infer` by instruction `callq`.
```
➜  release-centos7-llvm git:(pch) ✗ ./tiflash/tiflash version
TiFlash
Release Version: v6.2.0-alpha-30-gf2a406a
Edition:         Community
Git Commit Hash: f2a406aa93f06e98e1868597f0ff25ae87ed40c5
Git Branch:      pch
UTC Build Time:  2022-06-05 07:30:43
Enable Features: jemalloc avx avx512 unwind thinlto
Profile:         RELWITHDEBINFO

Raft Proxy
Git Commit Hash:   ca2f51f94e55bdd23749dcc02ab4afb94eeb5ae5
Git Commit Branch: HEAD
UTC Build Time:    2022-06-05 07:33:25
Rust Version:      rustc 1.60.0-nightly (1e12aef3f 2022-02-13)
Storage Engine:    tiflash
Prometheus Prefix: tiflash_proxy_
Profile:           release

➜  release-centos7-llvm git:(pch) ✗ objdump -d ./tiflash/tiflash | grep 'PlusDecimalInferer'
0000000001ce43c0 <_ZN2DB18PlusDecimalInferer5inferEjjjj>:```

➜  release-centos7-llvm git:(pch) ✗ objdump -d ./tiflash/tiflash | grep '1ce43c0'           
0000000001ce43c0 <_ZN2DB18PlusDecimalInferer5inferEjjjj>:
 1ce43c0:       55                      push   %rbp

**Review by fuzhe1989 — APPROVED** [review]

File: dbms/src/Common/Decimal.h (hunk 1 above)
Comment (solotzg) [reply dbms/src/Common/Decimal.h]: Although I think there is no risk, but... follow reviewer's idea.

**Pull request merged**